numpy/_core/src/umath/string_buffer.h

```diff
@@ -390,6 +494,90 @@ struct Buffer {
         return unary_loop<IMPLEMENTED_UNARY_FUNCTIONS::ISDIGIT>();
     }
 
+    inline bool
+    isalnum()
+    {
+        return unary_loop<IMPLEMENTED_UNARY_FUNCTIONS::ISALNUM>();
+    }
+
+    inline bool
+    islower()
+    {
+        size_t len = num_codepoints();
+        if (len == 0) {
+            return false;
+        }
+
+        Buffer<enc> tmp = *this;
+        bool cased = 0;
+        for (size_t i = 0; i < len; i++) {
+            if (codepoint_isupper<enc>(*tmp) || codepoint_istitle<enc>(*tmp)) {
+                return false;
+            }
+            else if (!cased && codepoint_islower<enc>(*tmp)) {
+                cased = true;
+            }
+            tmp++;
+        }
+        return cased;
+    }
+
+    inline bool
+    isupper()
+    {
+        size_t len = num_codepoints();
+        if (len == 0) {
+            return false;
+        }
+
+        Buffer<enc> tmp = *this;
+        bool cased = 0;
+        for (size_t i = 0; i < len; i++) {
+            if (codepoint_islower<enc>(*tmp) || codepoint_istitle<enc>(*tmp)) {
+                return false;
+            }
+            else if (!cased && codepoint_isupper<enc>(*tmp)) {
+                cased = true;
+            }
+            tmp++;
+        }
+        return cased;
+    }
+
+    inline bool
+    istitle()
+    {
+        size_t len = num_codepoints();
+        if (len == 0) {
+            return false;
+        }
+
+        Buffer<enc> tmp = *this;
+        bool cased = false;
+        bool previous_is_cased = false;
+        for (size_t i = 0; i < len; i++) {
+            if (codepoint_isupper<enc>(*tmp) || codepoint_istitle<enc>(*tmp)) {
+                if (previous_is_cased) {
+                    return false;
+                }
+                previous_is_cased = true;
+                cased = true;
+            }
+            else if (codepoint_islower<enc>(*tmp)) {
+                if (!previous_is_cased) {
+                    return false;
+                }
+                previous_is_cased = true;
```

Comment:
This one is not needed, since we know `previous_is_cased == true` - otherwise you would have returned.